src/derive/field/tower.rs

```diff
@@ -0,0 +1,500 @@
+            pub fn to_bytes(&self) -> [u8; $field::SIZE * 2] {
+                let mut res = [0u8; $field::SIZE * 2];
+                let c0_bytes = self.c0.to_bytes();
+                let c1_bytes = self.c1.to_bytes();
+                res[0..$field::SIZE].copy_from_slice(&c0_bytes[..]);
+                res[$field::SIZE..$field::SIZE * 2].copy_from_slice(&c1_bytes[..]);
+                res
+            }
+        }
+
+        impl ff::Field for $tower {
+            const ZERO: Self = Self::zero();
+            const ONE: Self = Self::one();
+
+            fn random(mut rng: impl RngCore) -> Self {
+                $tower {
+                    c0: $field::random(&mut rng),
+                    c1: $field::random(&mut rng),
+                }
+            }
+
+            fn is_zero(&self) -> Choice {
+                self.c0.is_zero() & self.c1.is_zero()
+            }
+
+            fn square(&self) -> Self {
+                self.square()
+            }
+
+            fn double(&self) -> Self {
+                self.double()
+            }
+
+            fn sqrt(&self) -> CtOption<Self> {
+                self.sqrt()
+            }
+
+            fn sqrt_ratio(num: &Self, div: &Self) -> (Choice, Self) {
+                ff::helpers::sqrt_ratio_generic(num, div)
+            }
+
+            fn invert(&self) -> CtOption<Self> {
+                self.invert()
+            }
+        }
+
+        impl PrimeField for $tower {
+            type Repr = $repr;
+
+            const MODULUS: &'static str = <$field as PrimeField>::MODULUS;
+            const MULTIPLICATIVE_GENERATOR: Self = $tower {
+                c0: $field::MULTIPLICATIVE_GENERATOR,
+                c1: $field::ZERO,
+            };
+            const NUM_BITS: u32 = $field::NUM_BITS;
+            const CAPACITY: u32 = $field::NUM_BITS;
+            const S: u32 = $field::S;
+
+            // TODO: Check that we can just 0 this and forget.
+            const ROOT_OF_UNITY: Self = $tower::ZERO;
+            const ROOT_OF_UNITY_INV: Self = $tower::ZERO;
+            const DELTA: Self = $tower::ZERO;
+
+            const TWO_INV: Self = $tower {
+                c0: $field::TWO_INV,
+                c1: $field::ZERO,
+            };
+
+            fn from_repr(repr: Self::Repr) -> CtOption<Self> {
+                let c0 = $field::from_bytes(&repr.0[..$field::SIZE].try_into().unwrap());
+                let c1 = $field::from_bytes(&repr.0[$field::SIZE..].try_into().unwrap());
+                CtOption::new($tower::new(c0.unwrap(), c1.unwrap()), Choice::from(1))
+            }
+
+            fn to_repr(&self) -> Self::Repr {
+                $repr(self.to_bytes())
+            }
+
+            fn is_odd(&self) -> Choice {
+                Choice::from(self.to_repr().as_ref()[0] & 1)
+            }
+        }
+
+        impl WithSmallOrderMulGroup<3> for $tower {
+            // $field::ZETA ^2
+            const ZETA: Self = $tower {
+                c0: ZETA,
```

Comment:
We could remove the `ZETA` constant in `fq2.rs` by using the generated one?

```suggestion
                c0: $field::ZETA,
```

Reply:
That's true. Same zeta is also a zeta in extension field. However I think we should use same scalar field zeta for G1 and G2 for example in BN to be consistent in `zeta_scalar * point = (zeta_base *x , y)`. We have to use other zeta for base field which is square of `zeta_base` so that  equation above can hold.

Reply:
I added this square calculation to compile time now

Reply:
Ah I see